pkg/storage/utils/decomposedfs/decomposed.go

```diff
@@ -0,0 +1,437 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package decomposedfs
+
+//go:generate mockery -name PermissionsChecker
+//go:generate mockery -name Tree
+
+import (
+	"context"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/logger"
+	"github.com/cs3org/reva/pkg/storage"
+	"github.com/cs3org/reva/pkg/storage/utils/chunking"
+	"github.com/cs3org/reva/pkg/storage/utils/decomposedfs/node"
+	"github.com/cs3org/reva/pkg/storage/utils/decomposedfs/options"
+	"github.com/cs3org/reva/pkg/storage/utils/decomposedfs/tree"
+	"github.com/cs3org/reva/pkg/storage/utils/decomposedfs/xattrs"
+	"github.com/cs3org/reva/pkg/storage/utils/templates"
+	"github.com/cs3org/reva/pkg/user"
+	"github.com/pkg/errors"
+	"github.com/pkg/xattr"
+)
+
+// PermissionsChecker defines an interface for checking permissions on a Node
+type PermissionsChecker interface {
+	AssemblePermissions(ctx context.Context, n *node.Node) (ap *provider.ResourcePermissions, err error)
+	HasPermission(ctx context.Context, n *node.Node, check func(*provider.ResourcePermissions) bool) (can bool, err error)
+}
+
+// Tree is used to manage a tree hierarchy
+type Tree interface {
+	Setup(owner string) error
+
+	GetMD(ctx context.Context, node *node.Node) (os.FileInfo, error)
+	ListFolder(ctx context.Context, node *node.Node) ([]*node.Node, error)
+	//CreateHome(owner *userpb.UserId) (n *node.Node, err error)
+	CreateDir(ctx context.Context, node *node.Node) (err error)
+	//CreateReference(ctx context.Context, node *node.Node, targetURI *url.URL) error
+	Move(ctx context.Context, oldNode *node.Node, newNode *node.Node) (err error)
+	Delete(ctx context.Context, node *node.Node) (err error)
+	RestoreRecycleItemFunc(ctx context.Context, key string) (*node.Node, func() error, error)
+	PurgeRecycleItemFunc(ctx context.Context, key string) (*node.Node, func() error, error)
+
+	WriteBlob(key string, reader io.Reader) error
+	ReadBlob(key string) (io.ReadCloser, error)
+	DeleteBlob(key string) error
+
+	Propagate(ctx context.Context, node *node.Node) (err error)
+}
+
+// Decomposedfs provides
+type Decomposedfs struct {
+	lu           *Lookup
+	tp           Tree
+	o            *options.Options
+	p            PermissionsChecker
+	chunkHandler *chunking.ChunkHandler
+}
+
+// NewDefault returns an instance with default components
+func NewDefault(m map[string]interface{}, bs tree.Blobstore) (storage.FS, error) {
```

Comment:
This would be slightly easier to read if values like `Lookup` and `Tree` are declared as package global variables, as seen [here](https://github.com/asim/go-micro/blob/d94936f6c97c5b1c40d179b55d9458419464b0e3/network/router/options.go#L77-L85) for example.